### DISCUSSION TITLE
transform algorithms written in Python from scikit-learn and Matplotlib with pyccel to c language

### DISCUSSION THREAD

**Discussion by madou-sow** [pyccel/pyccel#1039] [Q&A] [answered]
Hello ,

I would like to transform this application of python language into c language with your pyccel software. I already know that pyccel does not support Matplotlib and Scikit-learn at the moment. Is it possible to rewrite the program in python with other modules that pyccel supports?
Thank you in advance for the proposals

== program.py ==
```
print(__doc__)

import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import IsolationForest

rng = np.random.RandomState(42)

# Generate train data
X = 0.3 * rng.randn(100, 2)
X_train = np.r_[X + 2, X - 2]
# Generate some regular novel observations
X = 0.3 * rng.randn(20, 2)
X_test = np.r_[X + 2, X - 2]
# Generate some abnormal novel observations
X_outliers = rng.uniform(low=-4, high=4, size=(20, 2))

# fit the model
clf = IsolationForest(max_samples=100, random_state=rng)
clf.fit(X_train)
y_pred_train = clf.predict(X_train)
y_pred_test = clf.predict(X_test)
y_pred_outliers = clf.predict(X_outliers)

# plot the line, the samples, and the nearest vectors to the plane
xx, yy = np.meshgrid(np.linspace(-5, 5, 50), np.linspace(-5, 5, 50))
Z = clf.decision_function(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

plt.title("IsolationForest")
plt.contourf(xx, yy, Z, cmap=plt.cm.Blues_r)

b1 = plt.scatter(X_train[:, 0], X_train[:, 1], c='white',
                 s=20, edgecolor='k')
b2 = plt.scatter(X_test[:, 0], X_test[:, 1], c='green',
                 s=20, edgecolor='k')
c = plt.scatter(X_outliers[:, 0], X_outliers[:, 1], c='red',
                s=20, edgecolor='k')
plt.axis('tight')
plt.xlim((-5, 5))
plt.ylim((-5, 5))
plt.legend([b1, b2, c],
           ["training observations",
            "new regular observations", "new abnormal observations"],
           loc="upper left")
plt.show()
```
====
1- pyccel program.py --language=c

ERROR at parsing (syntax) stage
pyccel:
 |error [syntax]: pyplot.py [29]| Uncovered try/except/finally statements by Pyccel
 |fatal [syntax]: pyplot.py [75,13]| Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language (Dict(keys=[Attribute(value=Name(id='_api', ctx=Load()), attr='rename_parameter', ctx=Load()), Attribute(value=Name(id='_api', ctx=Load()), attr='make_keyword_only', ctx=Load())], values=[Attribute(value=Call(func=Attribute(value=Name(id='_api', ctx=Load()), attr='rename_parameter', ctx=Load()), args=[Constant(value='', kind=None), Constant(value='old', kind=None), Constant(value='new', kind=None), Lambda(args=arguments(posonlyargs=[], args=[arg(arg='new', annotation=None, type_comment=None)], vararg=None, kwonlyargs=[], kw_defaults=[], kwarg=None, defaults=[]), body=Constant(value=None, kind=None))], keywords=[]), attr='__code__', ctx=Load()), Attribute(value=Call(func=Attribute(value=Name(id='_api', ctx=Load()), attr='make_keyword_only', ctx=Load()), args=[Constant(value='', kind=None), Constant(value='p', kind=None), Lambda(args=arguments(posonlyargs=[], args=[arg(arg='p', annotation=None, type_comment=None)], vararg=None, kwonlyargs=[], kw_defaults=[], kwarg=None, defaults=[]), body=Constant(value=None, kind=None))], keywords=[]), attr='__code__', ctx=Load())]))

2- Can you give me the reference of all the classes and functions of pyccel indeed the link on the API reference of Pyccel ?

**Comment by EmilyBourne** [comment, marked as answer]
Hi,

Regarding your questions.
> Is it possible to rewrite the program in python with other modules that pyccel supports?

It is not currently possible and I don't think it will ever become possible to rewrite matplotlib with pyccel. But matplotlib is rarely the bottleneck of an application so we don't see this as a problem.
I had a quick look at scikit-learn, and it seems that it is written in python, so this may be possible. But it sounds like a lot of work to extract the functions and ensure that they are only using concepts that are already supported by pyccel. The scikit-learn code is full of classes, so it cannot currently be translated as is. You would have to write your own functions.

> Can you give me the reference of all the classes and functions of pyccel indeed the link on the API reference of Pyccel ?

We do not yet have an API reference. All available documentation is in the [tutorials](https://github.com/pyccel/pyccel/tree/master/tutorial) folder. The errors above are due to try/except (which does not exist in C/Fortran so is very difficult to translate) and `dict`. It would be nice to add `dict` to Pyccel, but we are not sure what the equivalent object(s) would be in C/Fortran currently
